crates/fixedpointmath/src/lib.rs

```diff
@@ -1,422 +1,406 @@
 mod macros;
+mod sign;
 mod utils;
+mod value;
 
+use core::panic;
 use std::{
     fmt,
-    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Shr, Sub, SubAssign},
-    str::FromStr,
+    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+    os::macos::raw,
 };
 
-use ethers::types::{Sign, I256, U256};
-use eyre::{eyre, ErrReport, Error, Result};
+use ethers::types::{I256, U256, U512};
+use eyre::{bail, eyre, Error, Report, Result};
 use rand::{
     distributions::{
-        uniform::{SampleBorrow, SampleUniform, UniformSampler},
+        uniform::{SampleBorrow, SampleUniform, UniformFloat, UniformSampler},
         Distribution, Standard,
     },
     Rng,
 };
+pub use sign::*;
 pub use utils::*;
+pub use value::*;
 
-/// A fixed point wrapper around the `U256` type from ethers-rs.
+/// A fixed point wrapper around ethers-rs.
+///
+/// This fixed point type is a heavily based on Solidity's FixedPointMath
+/// library with a few changes:
+/// - Support for negative numbers.
+/// - Support for overflowing intermediate operations in `mul_div_down` and
```

Comment:
Is this on by default? This seems like an anti-pattern for use-cases where we are trying to achieve parity

Reply:
Yes, we already reached parity, but now I'm going past it to address pitfalls in the solidity lib. Then we'll switch to using `FixedPoint<u128>` and `FixedPoint<i128>` in all the tests so that we can be sure the inputs are bounded by those integer types which we know are within ranges the Solidity can support. This way we stop running into errors related to bad inputs and focus on protocol correctness.

Reply:
Is the idea that the inputs are u128s but that the intermediate multiplications are done as U256? That makes sense to me. 
If so, this isn’t really a problem with the Solidity library. It’s more an issue with the sampling that we’re currently doing.

Reply:
Yeah I realized that as I was half way into the refactor. That this is really just a sampling issue, but I was already deep into the refactor and this will make the lib more flexible beyond parity with the lib used in Hyperdrive.